planner/cascades/stringer.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cascades
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb/planner/memo"
+)
+
+// ToString stringifies a Group Tree.
+func ToString(g *memo.Group) []string {
+	idMap := make(map[*memo.Group]int)
+	idMap[g] = 0
+	return toString(g, idMap, map[*memo.Group]struct{}{}, []string{})
+}
+
+// toString recursively stringifies a Group Tree using a preorder traversal method.
+func toString(g *memo.Group, idMap map[*memo.Group]int, visited map[*memo.Group]struct{}, strs []string) []string {
+	if _, exists := visited[g]; exists {
+		return strs
+	}
+	visited[g] = struct{}{}
+	// Add new Groups to idMap.
+	for item := g.Equivalents.Front(); item != nil; item = item.Next() {
+		expr := item.Value.(*memo.GroupExpr)
+		for _, childGroup := range expr.Children {
+			if _, exists := idMap[childGroup]; !exists {
+				idMap[childGroup] = len(idMap)
+			}
+		}
+	}
+	// Visit self first.
+	strs = append(strs, groupToString(g, idMap)...)
+	// Visit children then.
+	for item := g.Equivalents.Front(); item != nil; item = item.Next() {
+		expr := item.Value.(*memo.GroupExpr)
+		for _, childGroup := range expr.Children {
+			strs = toString(childGroup, idMap, visited, strs)
+		}
+	}
+	return strs
+}
+
+// groupToString only stringifies a single Group.
+// Format:
+// Group#1 Column: [Column#1,Column#2,Column#13] Unique key: []
+//     Selection_4 input:[Group#2], eq(Column#13, Column#2), gt(Column#1, 10)
+//     Projection_15 input:Group#3 Column#1, Column#2
+func groupToString(g *memo.Group, idMap map[*memo.Group]int) []string {
+	schema := g.Prop.Schema
+	colStrs := make([]string, 0, len(schema.Columns))
+	for _, col := range schema.Columns {
+		colStrs = append(colStrs, col.String())
+	}
+
+	groupLine := bytes.NewBufferString("")
+	fmt.Fprintf(groupLine, "Group#%d Schema:[%s]", idMap[g], strings.Join(colStrs, ","))
+
+	if len(g.Prop.Schema.Keys) > 0 {
+		ukStrs := make([]string, 0, len(schema.Keys))
+		for _, key := range schema.Keys {
+			ukColStrs := make([]string, 0, len(key))
+			for _, col := range key {
+				ukColStrs = append(ukColStrs, col.String())
+			}
+			ukStrs = append(ukStrs, "["+strings.Join(ukColStrs, ",")+"]")
+		}
+		fmt.Fprintf(groupLine, ", UniqueKey:[%s]", strings.Join(ukStrs, ","))
```

Comment:
seems this is not covered by the test?

Reply:
I export the function `buildKeyInfo` for this test, PTAL.

I will move `PruneColumn` and `BuildKeyInfo` into a preprocess phase of the cascades planner in next PR.